Modules/_winapi.c

```diff
@@ -774,12 +774,172 @@ gethandle(PyObject* obj, const char* name)
     return ret;
 }
 
+static PyObject*
+sortenvironmentkey(PyObject *module, PyObject *item)
+{
+    return _winapi_LCMapStringEx_impl(NULL, LOCALE_NAME_INVARIANT,
+                                      LCMAP_UPPERCASE, item);
+}
+
+static PyMethodDef sortenvironmentkey_def = {
+    "sortenvironmentkey",
+    _PyCFunction_CAST(sortenvironmentkey),
+    METH_O,
+    "",
+};
+
+bool
+sort_environment_keys(PyObject *keys)
+{
+    PyObject *keyfunc = NULL, *kwnames = NULL;
+    bool result = false;
+
+    keyfunc = PyCFunction_New(&sortenvironmentkey_def, NULL);
+    if (keyfunc == NULL) {
+        return false;
+    }
+    kwnames = Py_BuildValue("(s)", "key");
+    if (kwnames == NULL) {
+        goto cleanup;
+    }
+    PyObject *args[] = { keys, keyfunc };
+    PyObject *ret = PyObject_VectorcallMethod(&_Py_ID(sort), args, 1, kwnames);
+    if (ret == NULL) {
+        goto cleanup;
+    }
+    Py_DECREF(ret);
+
+    result = true;
+
+cleanup:
+    Py_XDECREF(keyfunc);
+    Py_XDECREF(kwnames);
+
+    return result;
+}
+
+static PyObject *
+dedup_environment_keys(PyObject *keys)
+{
+    PyObject *result = NULL;
+    wchar_t *prev_key_string = NULL;
+
+    result = PyList_New(0);
+    if (result == NULL) {
+        return NULL;
+    }
+    for (Py_ssize_t i=PyList_GET_SIZE(keys)-1; i>=0; i--) {
+        PyObject *key = PyList_GET_ITEM(keys, i);
+        if (i == PyList_GET_SIZE(keys)) {
+            prev_key_string = PyUnicode_AsWideCharString(key, NULL);
+            if (prev_key_string == NULL) {
+                goto error;
+            }
+            if (PyList_Insert(result, 0, key) < 0) {
+                goto error;
+            }
+            continue;
+        }
+        wchar_t *key_string = PyUnicode_AsWideCharString(key, NULL);
+        if (key_string == NULL) {
+            goto error;
+        }
+        if (CompareStringOrdinal(prev_key_string, -1, key_string, -1, TRUE) == CSTR_EQUAL) {
+            PyMem_Free(key_string);
+            continue;
+        }
+        if (PyList_Insert(result, 0, key) < 0) {
+            goto error;
+        }
+        PyMem_Free(prev_key_string);
+        prev_key_string = key_string;
```

Comment:
The memory resource handling is not easy to follow; I would prefer if it was more explicit. AFAICS, `prev_key_string` leaks.

Reply:
I found a way to get rid of the temporary `prev_key_string`, and added some comments on it: https://github.com/aisk/cpython/blob/d08808397a6c224c976aca7ee8def901606b52be/Modules/_winapi.c#L841-L844

Furthermore, I broke the string comparison parts into a separate function.

Reply:
Thanks, the resulting code is much more readable!